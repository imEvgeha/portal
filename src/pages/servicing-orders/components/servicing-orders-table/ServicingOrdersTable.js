import {camelCase, startCase} from 'lodash';
import React, {useEffect, useState} from 'react';
import {compose} from 'redux';
import EmphasizedCellRenderer from '../../../../ui/elements/nexus-grid/elements/cell-renderer/emphasized-cell-renderer/EmphasizedCellRenderer';
import withFilterableColumns from '../../../../ui/elements/nexus-grid/hoc/withFilterableColumns';
import withInfiniteScrolling from '../../../../ui/elements/nexus-grid/hoc/withInfiniteScrolling';
import withSideBar from '../../../../ui/elements/nexus-grid/hoc/withSideBar';
import NexusGrid from '../../../../ui/elements/nexus-grid/NexusGrid';
import {DATETIME_FIELDS} from '../../../../util/date-time/constants';
import {ISODateToView} from '../../../../util/date-time/DateTimeUtils';
import columnDefs from '../../columnMappings.json';
import {servicingOrdersService} from '../../servicingOrdersService';
import ServicingOrdersTableStatusBar from '../servicing-orders-table-status-bar/ServicingOrdersTableStatusBar';
import './ServicingOrdersTable.scss';

const ServicingOrderGrid = compose(
    withSideBar(),
    withFilterableColumns(),
    withInfiniteScrolling({fetchData: servicingOrdersService.getServicingOrders})
)(NexusGrid);

const ServicingOrdersTable = ({fixedFilter, externalFilter, setSelectedServicingOrders, refreshData, dataRefreshComplete}) => {
    const [statusBarInfo, setStatusBarInfo] = useState({
        totalRows: 0,
        selectedRows: 0,
    });
    const [gridApi, setGridApi] = useState(null);

    const valueFormatter = ({dataType = '', field = '', isEmphasized = false}) => {
        switch (dataType) {
            case 'string':
                return params => {
                    const {data = {}} = params || {};
                    const {[field]: value = ''} = data || {};
                    // Capitalizes every word and removes non-alphanumeric characters if string is emphasized
                    return isEmphasized ? startCase(camelCase(value)) : value;
                };
            case DATETIME_FIELDS.REGIONAL_MIDNIGHT:
                return params => {
                    const {data = {}} = params || {};
                    const {[field]: date = ''} = data || {};
                    return ISODateToView(date, dataType);
                };
        }
    };

    const updateColumnDefs = columnDefs => {
        return columnDefs.map(columnDef => ({
            ...columnDef,
            valueFormatter: valueFormatter(columnDef),
            cellRenderer: columnDef.isEmphasized ? 'emphasizedStringCellRenderer' : 'loadingCellRenderer',
        }));
    };

    /**
     * Callback when datatable is first rendered on the DOM
     * @param {object} param
     */
    const onFirstDataRendered = ({api}) => {
        // Resizes the table to fit the current width when first rendered.
        // Needs to be removed if more table rows are added to prevent overcrowding
        api.sizeColumnsToFit();
        setGridApi(api);
    };

    /**
     * Runs when the selections are changed.
     * This current function is being used to gather an array of so_numbers of the selected rows
     * @param params - the grid params object containing the api
     */
    const onSelectionChanged = ({api}) => {
        const selectedRowsData = api.getSelectedNodes().map(node => node.data);

        // set the new array to state
        setSelectedServicingOrders(selectedRowsData);
        setStatusBarInfo({...statusBarInfo, selectedRows: selectedRowsData.length});
    };

    const setTotalCount = total => {
        setStatusBarInfo({...statusBarInfo, totalRows: total});
    };

    const [columns, setColumns] = useState(updateColumnDefs(columnDefs));

    useEffect(
        () => {
            setColumns(updateColumnDefs(columnDefs));
        },
        [columnDefs]
    );

    useEffect(
        () => {
            if (refreshData) {
                // Refresh data
                gridApi.purgeInfiniteCache();

                // Remove all selections
                gridApi.deselectAll();
                setSelectedServicingOrders([]);

                dataRefreshComplete();
            }
        },
        [refreshData]
    );

    return (
        <div className="nexus-c-servicing-orders-table">
            <ServicingOrderGrid
                columnDefs={columns}
                mapping={columnDefs}
                frameworkComponents={{
                    emphasizedStringCellRenderer: EmphasizedCellRenderer,
                }}
                fixedFilter={fixedFilter}
                externalFilter={externalFilter}
                onFirstDataRendered={onFirstDataRendered}
                customDateFilterParamSuffixes={['Start', 'End']}
                onSelectionChanged={onSelectionChanged}
                rowSelection="multiple"
                // lets users deselect a row with cmd/ctrl + click
                rowDeselection={true}
                setTotalCount={setTotalCount}
            />
            <ServicingOrdersTableStatusBar statusBarInfo={statusBarInfo} />
        </div>
    );
};

export default ServicingOrdersTable;
