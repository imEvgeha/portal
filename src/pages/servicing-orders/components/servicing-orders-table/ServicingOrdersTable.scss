@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/mixins';
@import '~@vubiquity-nexus/portal-styles/scss/config';

.nexus-c-servicing-orders-table {
    max-width: 100%;
    height: calc(100vh - 205px);
    position: relative;

    .nexus-c-nexus-grid {
        height: 100%;
    }
    .ag-react-container a {
        color: map-get($colors, nexus-link);
        text-decoration: underline;
    }

    &__checkbox-cell {
        @include addSpacing('width', 11.25);
        margin-left: 0 !important;
        padding-left: 4px !important;
        display: flex;
        justify-content: flex-start;
    }

    &__checkbox--disabled {
        @include addSpacing('height', 3.25);
        @include addSpacing('width', 3.25);
        @include margin(1.5, 0, 0, 1.5);
        @include addSpacing('border-radius', 0.25);
        background-color: map-get($mid-neutral-colors, mid-neutral-8);
    }
}
