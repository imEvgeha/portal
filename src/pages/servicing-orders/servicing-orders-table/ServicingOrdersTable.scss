.nexus-c-servicing-orders-table {
  width: 100%;
  height: 100%;
  padding: 24px 24px 45px 24px;

  & .nexus-c-nexus-grid {
    height: 100%;
  }
}