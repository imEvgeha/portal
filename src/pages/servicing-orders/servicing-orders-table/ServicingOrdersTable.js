import React from 'react';
import {camelCase, startCase} from 'lodash';
import {compose} from 'redux';
import {servicingOrdersService} from './servicingOrdersService';
import NexusGrid from '../../../ui/elements/nexus-grid/NexusGrid';
import withInfiniteScrolling from '../../../ui/elements/nexus-grid/hoc/withInfiniteScrolling';
import withSideBar from '../../../ui/elements/nexus-grid/hoc/withSideBar';
import EmphasizedCellRenderer
    from '../../../ui/elements/nexus-grid/elements/cell-renderer/emphasized-cell-renderer/EmphasizedCellRenderer';
import columnDefs from '../../../../profile/servicingOrdersMapping.json';
import {DATETIME_FIELDS, ISODateToView} from '../../../util/DateTimeUtils';
import './ServicingOrdersTable.scss';

const ServicingOrderGrid = compose(
    withSideBar(),
    withInfiniteScrolling({fetchData: servicingOrdersService.getServicingOrders})
)(NexusGrid);

const ServicingOrdersTable = () => {
    const valueFormatter= ({dataType = '', field = '', isEmphasized = false}) => {
        switch (dataType) {
            case 'string':
                return (params) => {
                    const {data = {}} = params || {};
                    const {[field]: value = ''} = data || {};
                    // Capitalizes every word and removes non-alphanumeric characters if string is emphasized
                    return isEmphasized ? startCase(camelCase(value)) : value;
                };
            case DATETIME_FIELDS.REGIONAL_MIDNIGHT:
                return (params) => {
                    const {data = {}} = params || {};
                    const {[field]: date = ''} = data || {};
                    return ISODateToView(date, dataType);
                };
        }
    };

    const updateColumnDefs = (columnDefs) => {
        return columnDefs.map(columnDef => (
            {
                ...columnDef,
                valueFormatter: valueFormatter(columnDef),
                cellRenderer: (columnDef.isEmphasized)
                    ? 'emphasizedStringCellRenderer'
                    : 'loadingCellRenderer',
            }
        ));
    };

    return (
        <div className="nexus-c-servicing-orders-table">
            <ServicingOrderGrid
                columnDefs={updateColumnDefs(columnDefs)}
                frameworkComponents={{
                    emphasizedStringCellRenderer: EmphasizedCellRenderer,
                }}
            />
        </div>
    );
};

export default ServicingOrdersTable;
