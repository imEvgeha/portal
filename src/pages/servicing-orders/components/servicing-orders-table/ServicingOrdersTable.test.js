import React from 'react';
import {shallow} from 'enzyme';
import ServicingOrdersTable from './ServicingOrdersTable';

describe('ServicingOrdersTable', () => {
    let wrapper = null;

    beforeEach(() => {
        wrapper = shallow(<ServicingOrdersTable />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
