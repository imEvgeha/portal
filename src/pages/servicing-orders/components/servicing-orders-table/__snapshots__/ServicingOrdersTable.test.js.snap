// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServicingOrdersTable should match snapshot 1`] = `
<div
  className="nexus-c-servicing-orders-table"
>
  <ComposedComponent
    columnDefs={
      Array [
        Object {
          "cellRenderer": "loadingCellRenderer",
          "cellRendererParams": Object {
            "link": "/servicing-orders/",
            "newTab": false,
          },
          "colId": "so_number",
          "dataType": "string",
          "displayName": "SO Number",
          "enableSearch": true,
          "field": "so_number",
          "headerName": "SO Number",
          "javaVariableName": "so_number",
          "searchDataType": "string",
          "valueFormatter": [Function],
          "width": 150,
        },
        Object {
          "cellRenderer": "loadingCellRenderer",
          "colId": "description",
          "dataType": "string",
          "displayName": "SO Description",
          "enableSearch": true,
          "field": "description",
          "headerName": "SO Description",
          "javaVariableName": "description",
          "searchDataType": "string",
          "valueFormatter": [Function],
          "width": 200,
        },
        Object {
          "cellRenderer": "loadingCellRenderer",
          "colId": "external_id",
          "dataType": "string",
          "displayName": "SR Number",
          "enableSearch": true,
          "field": "external_id",
          "headerName": "SR Number",
          "javaVariableName": "external_id",
          "searchDataType": "string",
          "valueFormatter": [Function],
          "width": 200,
        },
        Object {
          "cellRenderer": "loadingCellRenderer",
          "colId": "sr_due_date",
          "dataType": "regionalMidnight",
          "displayName": "SR Due Date",
          "enableSearch": true,
          "field": "sr_due_date",
          "headerName": "SR Due Date",
          "javaVariableName": "sr_due_date",
          "searchDataType": "regionalMidnight",
          "valueFormatter": [Function],
          "width": 200,
        },
        Object {
          "cellRenderer": "loadingCellRenderer",
          "colId": "submitted_date",
          "dataType": "regionalMidnight",
          "displayName": "Submitted Date",
          "enableSearch": true,
          "field": "submitted_date",
          "headerName": "Submitted Date",
          "javaVariableName": "submitted_date",
          "searchDataType": "regionalMidnight",
          "valueFormatter": [Function],
          "width": 200,
        },
        Object {
          "cellRenderer": "loadingCellRenderer",
          "colId": "rush_order",
          "dataType": "string",
          "displayName": "Rush",
          "enableSearch": true,
          "field": "rush_order",
          "headerName": "Rush",
          "javaVariableName": "rush_order",
          "searchDataType": "string",
          "valueFormatter": [Function],
          "width": 100,
        },
        Object {
          "cellRenderer": "emphasizedStringCellRenderer",
          "colId": "status",
          "dataType": "select",
          "displayName": "Status",
          "enableSearch": true,
          "field": "status",
          "headerName": "Status",
          "isEmphasized": true,
          "javaVariableName": "status",
          "options": Array [
            "NOT_STARTED",
            "IN_PROGRESS",
            "COMPLETE",
            "CANCELLED",
            "FAILED",
          ],
          "searchDataType": "multiselect",
          "valueFormatter": undefined,
          "width": 150,
        },
        Object {
          "cellRenderer": "emphasizedStringCellRenderer",
          "colId": "readiness",
          "dataType": "select",
          "displayName": "Readiness",
          "enableSearch": true,
          "field": "readiness",
          "headerName": "Readiness",
          "isEmphasized": true,
          "javaVariableName": "readiness",
          "options": Array [
            "NEW",
            "READY",
            "ON_HOLD",
          ],
          "searchDataType": "multiselect",
          "valueFormatter": undefined,
          "width": 150,
        },
      ]
    }
    customDateFilterParamSuffixes={
      Array [
        "Start",
        "End",
      ]
    }
    frameworkComponents={
      Object {
        "emphasizedStringCellRenderer": [Function],
      }
    }
    mapping={
      Array [
        Object {
          "cellRendererParams": Object {
            "link": "/servicing-orders/",
            "newTab": false,
          },
          "colId": "so_number",
          "dataType": "string",
          "displayName": "SO Number",
          "enableSearch": true,
          "field": "so_number",
          "headerName": "SO Number",
          "javaVariableName": "so_number",
          "searchDataType": "string",
          "width": 150,
        },
        Object {
          "colId": "description",
          "dataType": "string",
          "displayName": "SO Description",
          "enableSearch": true,
          "field": "description",
          "headerName": "SO Description",
          "javaVariableName": "description",
          "searchDataType": "string",
          "width": 200,
        },
        Object {
          "colId": "external_id",
          "dataType": "string",
          "displayName": "SR Number",
          "enableSearch": true,
          "field": "external_id",
          "headerName": "SR Number",
          "javaVariableName": "external_id",
          "searchDataType": "string",
          "width": 200,
        },
        Object {
          "colId": "sr_due_date",
          "dataType": "regionalMidnight",
          "displayName": "SR Due Date",
          "enableSearch": true,
          "field": "sr_due_date",
          "headerName": "SR Due Date",
          "javaVariableName": "sr_due_date",
          "searchDataType": "regionalMidnight",
          "width": 200,
        },
        Object {
          "colId": "submitted_date",
          "dataType": "regionalMidnight",
          "displayName": "Submitted Date",
          "enableSearch": true,
          "field": "submitted_date",
          "headerName": "Submitted Date",
          "javaVariableName": "submitted_date",
          "searchDataType": "regionalMidnight",
          "width": 200,
        },
        Object {
          "colId": "rush_order",
          "dataType": "string",
          "displayName": "Rush",
          "enableSearch": true,
          "field": "rush_order",
          "headerName": "Rush",
          "javaVariableName": "rush_order",
          "searchDataType": "string",
          "width": 100,
        },
        Object {
          "colId": "status",
          "dataType": "select",
          "displayName": "Status",
          "enableSearch": true,
          "field": "status",
          "headerName": "Status",
          "isEmphasized": true,
          "javaVariableName": "status",
          "options": Array [
            "NOT_STARTED",
            "IN_PROGRESS",
            "COMPLETE",
            "CANCELLED",
            "FAILED",
          ],
          "searchDataType": "multiselect",
          "width": 150,
        },
        Object {
          "colId": "readiness",
          "dataType": "select",
          "displayName": "Readiness",
          "enableSearch": true,
          "field": "readiness",
          "headerName": "Readiness",
          "isEmphasized": true,
          "javaVariableName": "readiness",
          "options": Array [
            "NEW",
            "READY",
            "ON_HOLD",
          ],
          "searchDataType": "multiselect",
          "width": 150,
        },
      ]
    }
  />
</div>
`;
